File: SoftLogic.Core/Win/Security/Support/CustomPrincipal.cs
using System.Text.RegularExpressions;
using System.Diagnostics;
using System;
using System.Drawing.Drawing2D;
using System.Windows.Forms;
using System.Collections;
using System.Drawing;
using Microsoft.VisualBasic;
using System.Data.SqlClient;
using System.Data;
using System.Collections.Generic;
using WeifenLuo.WinFormsUI;
using Microsoft.Win32;
using WeifenLuo;
using System.Threading;
using System.Security.Principal;


namespace SoftLogik.Win
{
	namespace Security
	{
		internal class CustomPrincipal : IPrincipal
		{
			
			private IIdentity m_User;
			private IPrincipal m_OldPrincipal;
			private IUserManager m_UserManager;
			private string m_ApplicationName;
			private string[] m_Roles;
			private static bool m_ThreadPolicySet = false;
			
			private CustomPrincipal(IIdentity user, string applicationName, IUserManager userManager, bool cacheRoles)
			{
				m_OldPrincipal = Thread.CurrentPrincipal;
				
				m_User = user;
				m_ApplicationName = applicationName;
				m_UserManager = userManager;
				
				if (cacheRoles)
				{
					m_Roles = m_UserManager.GetRoles(m_ApplicationName, m_User.Name);
				}
				//Make this object the principal for this thread
				Thread.CurrentPrincipal = this;
			}
			public static void Attach(IIdentity user, string applicationName, IUserManager userManager)
			{
				Attach(user, applicationName, userManager, false);
			}
			public static void Attach(IIdentity user, string applicationName, IUserManager userManager, bool cacheRoles)
			{
				Debug.Assert(user.IsAuthenticated);
				
				IPrincipal _customPrincipal = new CustomPrincipal(user, applicationName, userManager, cacheRoles);
				
				//Make sure all future threads in this app domain use this principal
				//but because default principal cannot be set twice:
				if (m_ThreadPolicySet == false)
				{
					AppDomain currentDomain = AppDomain.CurrentDomain;
					currentDomain.SetThreadPrincipal(_customPrincipal);
					m_ThreadPolicySet = true;
				}
			}
			public void Detach()
			{
				Thread.CurrentPrincipal = m_OldPrincipal;
			}
			
			public IIdentity Identity
			{
				get
				{
					return m_User;
				}
			}
			public bool IsInRole(string role)
			{
				if (m_Roles != null)
				{
					foreach (string itm in m_Roles)
					{
						if (itm == role)
						{
							return true;
						}
					}
                    return false;
				}
				else
				{
					return m_UserManager.IsInRole(m_ApplicationName, m_User.Name, role);
				}
			}
			
		}
	}
	
}

File: SPCode/Security/Support/CustomPrincipal.vb
Imports System
Imports System.Threading
Imports System.Diagnostics
Imports System.Security.Principal

Namespace Security
    Friend Class CustomPrincipal
        Implements IPrincipal
        Private m_User As IIdentity
        Private m_OldPrincipal As IPrincipal
        Private m_UserManager As IUserManager
        Private m_ApplicationName As String
        Private m_Roles As String()
        Private Shared m_ThreadPolicySet As Boolean = False

        Private Sub New(ByVal user As IIdentity, ByVal applicationName As String, ByVal userManager As IUserManager, ByVal cacheRoles As Boolean)
            m_OldPrincipal = Thread.CurrentPrincipal

            m_User = user
            m_ApplicationName = applicationName
            m_UserManager = userManager

            If cacheRoles Then
                m_Roles = m_UserManager.GetRoles(m_ApplicationName, m_User.Name)
            End If
            'Make this object the principal for this thread
            Thread.CurrentPrincipal = Me
        End Sub
        Public Shared Sub Attach(ByVal user As IIdentity, ByVal applicationName As String, ByVal userManager As IUserManager)
            Attach(user, applicationName, userManager, False)
        End Sub
        Public Shared Sub Attach(ByVal user As IIdentity, ByVal applicationName As String, ByVal userManager As IUserManager, ByVal cacheRoles As Boolean)
            Debug.Assert(user.IsAuthenticated)

            Dim _customPrincipal As IPrincipal = New CustomPrincipal(user, applicationName, userManager, cacheRoles)

            'Make sure all future threads in this app domain use this principal
            'but because default principal cannot be set twice:
            If m_ThreadPolicySet = False Then
                Dim currentDomain As AppDomain = AppDomain.CurrentDomain
                currentDomain.SetThreadPrincipal(_customPrincipal)
                m_ThreadPolicySet = True
            End If
        End Sub
        Public Sub Detach()
            Thread.CurrentPrincipal = m_OldPrincipal
        End Sub

        Public ReadOnly Property Identity() As IIdentity Implements IPrincipal.Identity
            Get
                Return m_User
            End Get
        End Property
        Public Function IsInRole(ByVal role As String) As Boolean Implements IPrincipal.IsInRole
            If Not m_Roles Is Nothing Then
                For Each itm As String In m_Roles
                    If itm = role Then
                        Return True
                    End If
                Next
            Else
                Return m_UserManager.IsInRole(m_ApplicationName, m_User.Name, role)
            End If
        End Function

    End Class
End Namespace
